### ISSUE TITLE
[temp.names] p4 cannot interpret the  `>` that is nested in multiple levels

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#432]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

[temp.names] p4 says:  
> When parsing a template-argument-list, the **first non-nested >** is taken as the ending delimiter rather than a greater-than operator. Similarly, **the first non-nested >>** is treated as two consecutive but distinct > tokens, the first of which is taken as the end of the template-argument-list and completes the template-id.

Such two rules both mention **the first non-nested**, so it is unclear how the token will be interpreted if the `>` or `>>` are nested tokens. Does it mean they should be interpreted as a greater-than operator(for `>`) or something else? 

Consider this example:
````cpp
A<B<C<D>>> var;
````
Where `A`, `B`, and `C` are class templates, `D` is a type. In this case, according to the rule the outermost `>>` is the first non-nested `>>`, hence they are interpreted as the end of the template-argument-list, respectively. However, what about the third last `>`, It **is the nested one** according to the note: 
> 116) A > that encloses the type-id of a dynamic_cast, static_cast, reinterpret_cast or const_cast, or which encloses the template-arguments of a subsequent template-id, is considered nested for the purpose of this description.

There is no rule in this subclause that specifies how this token will be interpreted

### Suggested Resolution

The rule should recursively apply to arbitrary multiple levels of templated-id. `A<B<C<E<F>>>>`, in this case, the second last `>>` is a nested token, which also cannot be interpreted by the rule

**Comment by jensmaurer** [comment]
This is the wrong direction. The note talks about "subsequent template-id", which A is not. I think the rule is fine as-is.

**Comment by xmh0511** [comment]
> This is the wrong direction. The note talks about "subsequent template-id", which A is not. I think the rule is fine as-is.

Isn't `C<D>` in `A<B<C<D>>>` the subsequent template-id? According to the current wording, the token `>` that precedes `>>` is the nested one, which cannot be interpreted as the ending delimiter of the template-parameters-list of `C`.

**Comment by jensmaurer** [comment]
We're parsing inside-out and left-to-right. We first need to know where the template-argument-list of `C<D>` ends before we can do anything at all with the outer template-argument-lists.

**Comment by xmh0511** [comment]
> We're parsing inside-out and left-to-right. We first need to know where the template-argument-list of `C<D>` ends before we can do anything at all with the outer template-argument-lists.

However, this meaning isn't embodied in the current wording. Furthermore, how do we parse template-argument-list inside-out? I think we should first determine the bound of a `template-argument-list`, this means, we must first find out the start and end of the *template-argument-list* for the outermost *template-id*, and the inner one and so forth.

**Comment by jensmaurer** [comment]
> we must first find out the start and end bound of the template-argument-list for the outermost template-id, and continue to find out bounds of the inner template-id, and so forth.

No, that can't work.  `A<B<C> >` (note spaces). We need to determine that the second-to-last `>` ends the inner template-argument-list (and thus we consume it at that point) before we can continue to determine that the last `>` ends the outer template-argument-list.

**Comment by xmh0511** [comment]
> > we must first find out the start and end bound of the template-argument-list for the outermost template-id, and continue to find out bounds of the inner template-id, and so forth.
> 
> No, that can't work. `A<B<C> >` (note spaces). We need to determine that the second-to-last `>` ends the inner template-argument-list (and thus we consume it at that point) before we can continue to determine that the last `>` ends the outer template-argument-list. These are the basics of a recursive-decent parser, I believe.

I see your meaning, you meant we must find out the **paired** `>` corresponding to the recent `<` we encountered in the token sequence, and consuming that `>` for `<` will let us continue to find out the next paired `>` for the last recent `<`. However, how do we word them?

The current wording "non-nested" is obviously insufficient. Any `>` ending the *template-argument-list* of the inner *template-id* is nested by the outermost template-id, which is **not** "non-nested".

**Comment by jensmaurer** [comment]
"nested" wants to refer to round parentheses and similar.

CWG579
